frida/src/script.rs

```diff
@@ -238,6 +237,46 @@ impl<'a> Script<'a> {
         self.rpc_id_counter
     }
 
+    /// Access exported functions from a Frida script.
+    pub fn exports(&mut self, function_name: &str, args: Option<Value>) -> Result<Option<Value>> {
```

Comment:
I don't love this interface. 

Firstly, it should be `args: &Vec<Value>`, to make it easier to call with multiple arguments.

Secondly, I don't love that we are called exports, but we actually run the export. Perhaps this method should be called `call` or `run`?

Reply:
> Firstly, it should be args: &Vec<Value>, to make it easier to call with multiple arguments.

Yeah, at the beginning I was thinking about using  `Vec<Value>` but I though the parameters might look a bit messy. For example:

```rust
// Instead of:
script.exports("nIncrement", Some(json!([2])));
// It would be:
script.exports("nIncrement", Some(vec![Value::Number(2.into())]));

// Instead of:
script.exports("bye", Some(json!(["Potato"])));
// It would be:
script.exports("bye", Some(vec![Value::String("Potato".to_string())]));

// And this:
script.exports("sumVals", Some(json!([[1, 2, 3, 4], true])))
// will turn into this:
script.exports(
    "sumVals",
    Some(vec![
        Value::Array(vec![1.into(), 2.into(), 3.into(), 4.into()]),
        Value::Bool(true),
    ]),
)
```

If you are ok with that or have another suggestion, I will be happy to make the changes. 

> args: `&Vec<Value>`

If we don't use `Option<>` then how are we going to deal with `None` values? ` Value::Null`? empty vector?

> Secondly, I don't love that we are called exports, but we actually run the export. Perhaps this method should be called call or run?

I named it `exports` to align with the other bindings. I think maintaining consistency across bindings is important, especially for documentation purposes. Since there's limited documentation on frida-rust bindings, this consistency will allow people to refer to existing documentation, like -> https://frida.re/docs/functions/, and guess based on other documentations already out there.

Reply:
I'm honestly conflicted on both points.

@oleavr do you have an opinion?

Reply:
I'm honestly conflicted on both points.

@oleavr do you have an opinion?

Reply:
Hey guys, any updates on this?

Reply:
How about something like:

```rust
script.exports.call("nIncrement", Some(json!([2])));
```

And you might do:

```rust
let agent = script.exports;
agent.call("nIncrement", Some(json!([2])));
agent.call("bye", Some(json!(["Potato"])));
```

Reply:
Hi @oleavr ! what is `exports` in this case? I am guessing it will be a `struct` what information are we going to store in it?

Reply:
👋🏻 It would need the underlying `FridaScript` (holding its own ref to it I suppose), and state for pending requests, so `call` can be implemented. It would also have to subscribe to the `message` signal so it can pick out the RPC messages.

Reply:
I noticed that this way I would have to move the functions `post` and `list_exports`, from `script` into `exports`.
like so:
```rust
script.list_exports()
script.post()

// would become:
script.exports.list_exports() // we might need to change the name of this function if we do that?
script.exports.post()
```

Is this ok?

Reply:
why do we have to move them?

Reply:
I jumped to conclusion without checking thoroughly. I don't think there is a nice way to do `script.exports.call()`

For example. I would have to have `exports` to be an `Option` in order for me to reference it self :

```rust
pub struct Exports<'a> {
    script_ref: *mut Script<'a>,
}


impl<'a> Script<'a> {
    pub(crate) fn from_raw(script_ptr: *mut _FridaScript) -> Script<'a> {
        let mut script = Script {
            script_ptr,
            phantom: PhantomData,
            rpc_id_counter: 0,
            callback_handler: CallbackHandler::new(),
            exports: None,
        };

        let exports = Exports {
            script_ref: &mut script as *mut Script,
        };

        script.exports = Some(exports);

        script
    }
```

I was also trying something like this:
```rust 
pub struct Exports<'a> {
    script_ptr: *mut _FridaScript,
    phantom: PhantomData<&'a _FridaScript>,
}
```
and that's when I thought I needed to move `post` since it is used by `call`

Maybe I am overthinking this?

Reply:
Can you not just keep a copy of the raw `_FridaScript` pointer in both the `Script` and `Exports`? (and maybe increase the refcount?)

Reply:
I can but, `call` needs to call `post` which is an `impl` of `Script`, so I won't be able to call `post` from `Exports`. I guess I could re-implement `post` inside `call`

Reply:
> How about something like:
> 
> ```rust
> script.exports.call("nIncrement", Some(json!([2])));
> ```
> 
> And you might do:
> 
> ```rust
> let agent = script.exports;
> agent.call("nIncrement", Some(json!([2])));
> agent.call("bye", Some(json!(["Potato"])));
> ```

We can now call `.exports.call`